### PR TITLE
tls: Use OpenSSL default trusted CA list and allow configuration of the trusted CA path

### PR DESCRIPTION
**Pull request by PaulSD**
Please see the commit messages for additional explanation.  These are actually two separate changes - I would have split this into two separate pull requests except that the changes are adjacent in the source code, and the documentation for the changes is interdependent, so I couldn't split this into two pull requests without causing merge conflicts between them.

### CODE REVIEW

**Comment by jasnell** [comment]
@indutny ... thoughts on this one?

**Comment by indutny** [comment]
I'm -1 for using default CA list, and +1 for introducing new API ;)

**Comment by PaulSD** [comment]
Why -1 for the default CA list?  Would it be better if I kept the existing behavior as the default, and added some options to allow the user to switch to the OpenSSL CA list instead? (Perhaps a compile-time option to enable the OpenSSL CA list, and/or an environment variable which causes the OpenSSL CA list to be used at runtime?)  I'd like to have some ability to change the CA list without changing code...

**Comment by indutny** [comment]
I like the idea of having an option to use it, but using it by default doesn't seem a good idea to me. Many systems has old list of CAs, and we are trying to keep it up to date in node.js using Mozilla's list. cc @bnoordhuis

**Comment by PaulSD** [comment]
Ok, thanks.  I'll update the patch to make this an option rather than the default...

File: src/node_crypto.cc [hunk 1]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
 
   assert(sc->ca_store_ == NULL);
 
+  if (args.Length() == 2) {
+    char* caFile = NULL;
+    char* caPath = NULL;
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: Can you use `nullptr` here and snake_case the variable names?

File: src/node_crypto.cc [hunk 2]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
...
+    if (args[0]->IsString()) {
+      String::Utf8Value file(args[0]);
+      caFile = *file;
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: This is unsafe; you're assigning a pointer to memory that goes out of scope immediately afterwards.  Happens a few lines below as well.

File: src/node_crypto.cc [hunk 3]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
...
+    } else if (!args[0]->IsUndefined() && !args[0]->IsNull()) {
+      return sc->env()->ThrowTypeError("Bad parameter");
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: I'd move the type checking to JS land and just `CHECK(args[0]->IsString())` here.

File: src/node_crypto.cc [hunk 4]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
...
+    }
+    if (args[1]->IsString()) {
+      String::Utf8Value path(args[1]);
+      caPath = *path;
+    } else if (!args[1]->IsUndefined() && !args[1]->IsNull()) {
+      return sc->env()->ThrowTypeError("Bad parameter");
+    }
+    if (!SSL_CTX_load_verify_locations(sc->ctx_, caFile, caPath))
+      return sc->env()->ThrowTypeError("Error loading CA certificates from caFile or caPath");
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: Long line.

File: src/node_crypto.cc [hunk 5]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
...
+    return;
+  }
+
+  if (args.Length() != 0)
+    return sc->env()->ThrowTypeError("Bad parameter");
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: I'd drop this and put a `CHECK(args.Length() == 0 || args.Length() == 2)` at the top.

File: src/node_crypto.cc [hunk 6]
```diff
@@ -699,6 +699,32 @@ void SecureContext::AddRootCerts(const FunctionCallbackInfo<Value>& args) {
...
+
+  if (SSL_CTX_set_default_verify_paths(sc->ctx_))
+    return;
```
Comment (bnoordhuis) [inline src/node_crypto.cc]: I understand how this works but it's very implicit.  Can you add a comment explaining it and maybe move it into the `if` block a few lines up?

**Comment by bnoordhuis** [comment]
One caveat with this change: openssl lazy-loads the certificate files blocking I/O.  If the files are located on a slow file system (like NFS), it's going to block the node process for some time.

**Pull request closed without merging**